File: src/main/java/susc/s01/Data/Log/JoinLog/PlayerLogHandler.java
package susc.s01.Data.Log.JoinLog;

import org.bukkit.Bukkit;
import susc.s01.Data.Log.Handler;
import susc.s01.Data.Log.LogData;

import java.util.ArrayList;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class PlayerLogHandler implements Handler<UUID, PlayerLog> {
    private static class JoinLoggerHandlerHolder {
        private final static PlayerLogHandler INSTANCE = new PlayerLogHandler();
    }

    public static PlayerLogHandler getInstance() {
        return JoinLoggerHandlerHolder.INSTANCE;
    }

    private final ConcurrentHashMap<UUID, PlayerLog> joinLogData = LogData.getJoinLogDataInstance();

    @Override
    public synchronized void putData(PlayerLog playerLog) {
        joinLogData.put(playerLog.uuid(), playerLog);
    }

    @Override
    public synchronized PlayerLog getData(UUID uuid) {
        return joinLogData.get(uuid);
    }

    @Override
    public synchronized Boolean replaceData(PlayerLog playerLog) throws Exception {
        PlayerLog legacy = joinLogData.get(playerLog.uuid());
        if (legacy.equals(playerLog))
            throw new Exception("Failed Replace Object > " + playerLog.uuid());

        joinLogData.replace(playerLog.uuid(), playerLog);
        return true;
    }

    @Override
    public synchronized Boolean removeData(UUID uuid) throws Exception  {
        if (joinLogData.get(uuid) == null)
            throw new Exception("Non Exists JoinData > " + uuid);

        joinLogData.remove(uuid);
        return true;
    }

    @Override
    public ArrayList<PlayerLog> getAllUserTable() {
        Bukkit.getOnlinePlayers().forEach(player -> getData(player.getUniqueId()));
        return new ArrayList<>(new ConcurrentHashMap<>(this.joinLogData).values());
    }

    @Override
    public void updateAllUserData(ArrayList<PlayerLog> newUserData) {
        this.joinLogData.clear();

        if (newUserData == null)
            return;

        newUserData.forEach(this::putData);
    }
}

File: src/main/java/susc/s01/Data/Log/JoinLog/PlayerLogHandler.kt
package susc.s01.Data.Log.JoinLog

import org.bukkit.Bukkit
import org.bukkit.entity.Player
import susc.s01.Data.Handler
import susc.s01.Data.Log.LogData
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer

class PlayerLogHandler : Handler<UUID, PlayerLog?> {
    private object JoinLoggerHandlerHolder {
        val instance: PlayerLogHandler = PlayerLogHandler()
            get() = JoinLoggerHandlerHolder.INSTANCE
    }

    private val joinLogData: ConcurrentHashMap<UUID?, PlayerLog?>? = LogData.getJoinLogDataInstance()

    @Synchronized
    override fun putData(playerLog: PlayerLog?) {
        joinLogData!![playerLog!!.uuid] = playerLog
    }

    @Synchronized
    override fun getData(uuid: UUID): PlayerLog? {
        return joinLogData!![uuid]
    }

    @Synchronized
    @Throws(Exception::class)
    override fun replaceData(playerLog: PlayerLog?): Boolean {
        val legacy = joinLogData!![playerLog!!.uuid]
        if (legacy == playerLog) throw Exception("Failed Replace Object > " + playerLog.uuid)

        joinLogData.replace(playerLog.uuid, playerLog)
        return true
    }

    @Synchronized
    @Throws(Exception::class)
    override fun removeData(uuid: UUID): Boolean {
        if (joinLogData!![uuid] == null) throw Exception("Non Exists JoinData > $uuid")

        joinLogData.remove(uuid)
        return true
    }

    override val allUserTable: ArrayList<E>
        get() {
            Bukkit.getOnlinePlayers().forEach { player: Player -> getData(player.uniqueId) }
            return ArrayList(ConcurrentHashMap(this.joinLogData).values)
        }

    override fun updateAllUserData(newUserData: ArrayList<PlayerLog?>?) {
        joinLogData!!.clear()

        if (newUserData == null) return

        newUserData.forEach(Consumer { playerLog: PlayerLog? -> this.putData(playerLog) })
    }

    companion object {
    }
}
